### DISCUSSION TITLE
debounce

### DISCUSSION THREAD

**Discussion by xiao-ice** [hixb/my-express#4] [Q&A] [answered]
debounce

**Comment by hixb** [comment, marked as answer]
> debounce

`


const debounce = (func, time = 17, options = {
    leading: true,
    context: null
}) => {
    let timer;
    const _debounce = function (...args) {
        if (timer) {
            clearTimeout(timer)
        }
        if (options.leading && !timer) {
            timer = setTimeout(null, time)
            func.apply(options.context, args)
        }else{
            timer = setTimeout(() => {
                func.apply(options.context, args)
                timer = null
            }, time)
        }
    };

    _debounce.cancel = function () {
        clearTimeout(timer)
        timer = null
    };
    return _debounce
};


`
